tests/acceptance/17_users/unsafe/20_modify_user_missing_secondary_group.cf

```diff
@@ -0,0 +1,52 @@
+bundle common test_meta
+{
+  vars:
+    "description" string => "Managing a user without specifying any secondary groups attribute should result in no changes to existing secondary groups";
+    "store_id" string => "3710";
+    "covers" string => "operational_repaired";
+}
+
+body common control
+{
+      inputs => { "../../default.cf.sub", "user_queries.cf.sub" };
+      bundlesequence  => { default("$(this.promise_filename)") };
+}
+
+bundle agent init
+{
+  commands:
+    "/usr/bin/sudo $(user_tests.pw_path) userdel testu";
+    "/usr/bin/sudo $(user_tests.pw_path) delgroup testu";
+    "/usr/bin/sudo $(user_tests.pw_path) delgroup testg1";
+    "/usr/bin/sudo $(user_tests.pw_path) delgroup testg2";
+    "/usr/bin/sudo $(user_tests.pw_path) groupadd testg1";
+    "/usr/bin/sudo $(user_tests.pw_path) useradd testu";
+    freebsd::
+      "/usr/bin/sudo $(user_tests.pw_path) usermod testu --groups testg1 --append";
+    !freebsd::
+      "/usr/bin/sudo $(user_tests.pw_path) usermod --groups testg1 --append testu";
+}
+
+bundle agent test
+{
+  users:
+    "testu"
+      policy => "present";
+}
+
+bundle agent check
+{
+  methods:
+    "any" usebundle => user_exists("testu", "success", "failure"),
+      classes => always("methods_run");
```

Comment:
That would be the same effect right? @nickanderson where is "methods_run" defined? This seems to be a unique aspect of 17_users acceptance tests. Is it correct?

Reply:
@craigcomstock, `methods_run` is defined as a result of the methods promise.

The *always* classes body defines whatever is passed in as an argument no matter what the promise result is. It's basically the same as the `_reached` suffixed class that gets defined when using the *results* classes body

    body classes always(x)
    # @brief Define class `x` no matter what the outcome of the promise is
    # @param x The name of the class to be defined
    {
          promise_repaired => { "$(x)" };
          promise_kept => { "$(x)" };
          repair_failed => { "$(x)" };
          repair_denied => { "$(x)" };
          repair_timeout => { "$(x)" };
    }
    
    body classes results(scope, class_prefix)
    # @brief Define classes prefixed with `class_prefix` and suffixed with
    # appropriate outcomes: _kept, _repaired, _not_kept, _error, _failed,
    # _denied, _timeout, _reached
    #
    # @param scope The scope in which the class should be defined (`bundle` or `namespace`)
    # @param class_prefix The prefix for the classes defined
    #
    # This body can be applied to any promise and sets global
    # (`namespace`) or local (`bundle`) classes based on its outcome. For
    # instance, with `class_prefix` set to `abc`:
    #
    # * if the promise is to change a file's owner to `nick` and the file
    # was already owned by `nick`, the classes `abc_reached` and
    # `abc_kept` will be set.
    #
    # * if the promise is to change a file's owner to `nick` and the file
    # was owned by `adam` and the change succeeded, the classes
    # `abc_reached` and `abc_repaired` will be set.
    #
    # This body is a simpler, more consistent version of the body
    # `scoped_classes_generic`, which see. The key difference is that
    # fewer classes are defined, and only for outcomes that we can know.
    # For example this body does not define "OK/not OK" outcome classes,
    # since a promise can be both kept and failed at the same time.
    #
    # It's important to understand that promises may do multiple things,
    # so a promise is not simply "OK" or "not OK." The best way to
    # understand what will happen when your specific promises get this
    # body is to test it in all the possible combinations.
    #
    # **Suffix Notes:**
    #
    # * `_reached` indicates the promise was tried. Any outcome will result
    #   in a class with this suffix being defined.
    #
    # * `_kept` indicates some aspect of the promise was kept
    #
    # * `_repaired` indicates some aspect of the promise was repaired
    #
    # * `_not_kept` indicates some aspect of the promise was not kept.
    #   error, failed, denied and timeout outcomes will result in a class
    #   with this suffix being defined
    #
    # * `_error` indicates the promise repair encountered an error
    #
    # * `_failed` indicates the promise failed
    #
    # * `_denied` indicates the promise repair was denied
    #
    # * `_timeout` indicates the promise timed out
    #
    # **Example:**
    #
    # ```cf3
    # bundle agent example
    # {
    #   commands:
    #     "/bin/true"
    #       classes => results("bundle", "my_class_prefix");
    #
    #   reports:
    #     my_class_prefix_kept::
    #       "My promise was kept";
    #
    #     my_class_prefix_repaired::
    #       "My promise was repaired";
    # }
    # ```
    #
    # **See also:** `scope`, `scoped_classes_generic`, `classes_generic`
    {
      scope => "$(scope)";
    
      promise_kept => { "$(class_prefix)_reached",
                        "$(class_prefix)_kept" };
    
      promise_repaired => { "$(class_prefix)_reached",
                            "$(class_prefix)_repaired" };
    
      repair_failed => { "$(class_prefix)_reached",
                         "$(class_prefix)_error",
                         "$(class_prefix)_not_kept",
                         "$(class_prefix)_failed" };
    
      repair_denied => { "$(class_prefix)_reached",
                         "$(class_prefix)_error",
                         "$(class_prefix)_not_kept",
                         "$(class_prefix)_denied" };
    
      repair_timeout => { "$(class_prefix)_reached",
                          "$(class_prefix)_error",
                          "$(class_prefix)_not_kept",
                          "$(class_prefix)_timeout" };
    }

I think that the suggestion from @vpodzime to do `always( "ready" )` is fine and simplifies the policy slightly. If you really want to be sure that each of those methods promises have run before the reports, then I would suggest using different classes for each, then defining ready based on all of those being true.

    methods:
      "any" usebundle => user_is_in_secondary_group("testu", "testg1", "testg1_success", "testg1_failure"),
        classes => always("methods_run_1");
    
      "any" usebundle => user_is_in_secondary_group("testu", "testg2", "testg2_success", "testg2_failure"),
        classes => always("methods_run_2");
    
      "any" usebundle => user_exists("testu", "testu_success", "testu_failure"),
        classes => always("methods_run_3");
    
    classes:
      "ready" and => { "methods_run_1", "methods_run_2", "methods_run_3" };

But, that's a bit overkill given the knowledge of how things work.

Reply:
Oh. Ok. I can refactor this a bit and make sure the other tests are still correct. Which I bet they are. I succumbed to copy/pasta. :+1: